# Sprint 4 Project
## Exploratory Data Analysis
In this notebook, I will perform some exploratory data analysis on a dataset that I retrieved from Kaggle.com. I will use my findings and my code from this notebook to create and deploy a webapp to share my analysis with others.

### Importing libraries and data

In [1]:
# import libraries
import pandas as pd
import streamlit as st
import plotly_express as px

#load the data into a DataFrame using pandas
games_df = pd.read_csv('../games.csv')
games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  object 
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  object 
 6   winner          20058 non-null  object 
 7   increment_code  20058 non-null  object 
 8   white_id        20058 non-null  object 
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  object 
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  object 
 13  opening_eco     20058 non-null  object 
 14  opening_name    20058 non-null  object 
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


There are no missing values. Next, check for duplicate rows.

In [2]:
# checking for duplicate rows
games_df.duplicated().sum()

429

There are 429 duplicate rows, further investigation is required.

In [3]:
# filter for duplicated rows and review them
duplicate_games = games_df[games_df.duplicated()]
duplicate_games


id  rated    created_at  last_move_at  turns victory_status  \
950    XRuQPSzH   True  1.503580e+12  1.503580e+12     50         resign   
1794   x1Svcl0P   True  1.499820e+12  1.499830e+12     59           mate   
1830   wgIHzqDo   True  1.499710e+12  1.499710e+12    113           mate   
2554   JrFxlGZo  False  1.489330e+12  1.489330e+12     78           mate   
2556   ezKACesj  False  1.488840e+12  1.488840e+12     51           mate   
...         ...    ...           ...           ...    ...            ...   
20013  Uk0pZCRR   True  1.499532e+12  1.499541e+12     95         resign   
20014  ZhZm1ol9   True  1.498871e+12  1.498879e+12    109           draw   
20015  fpqjttDx   True  1.497663e+12  1.497664e+12     42         resign   
20016  lzPgCJCS   True  1.497660e+12  1.497663e+12    102         resign   
20017  p391pBJv   True  1.497647e+12  1.497654e+12     97         resign   

      winner increment_code         white_id  white_rating         black_id  \
950    black           10+0      chess-brahs          2036             cdvh   
1794   white           10+0  kachess10204542          1171  quantummequanic   
1830   white           10+0  quantummequanic          1403     barrow-wight   
2554   black           15+0          sp37578          1309       penguin100   
2556   white           15+0       penguin100          1393          sp37578   
...      ...            ...              ...           ...              ...   
20013  black          45+45        kferapont          2094   moistvonlipwig   
20014   draw          45+45   moistvonlipwig          2229              mn8   
20015  black           7+30   waltersafarian          2162   moistvonlipwig   
20016  black           7+30   moistvonlipwig          2236   waltersafarian   
20017  white          60+30        i9_4_2001          2268   moistvonlipwig   

       black_rating                                              moves  \
950            1706  Nf3 d5 g3 c5 Bg2 Nc6 d3 Nf6 O-O e5 Nc3 d4 Ne4 ...   
1794           1432  c4 e5 Nc3 Nf6 Nf3 Nc6 g3 d5 cxd5 Nxd5 Bg2 Be6 ...   
1830           1364  e4 e6 e5 Bc5 d4 Be7 c3 Nh6 Bxh6 gxh6 Qd2 Bg5 f...   
2554           1393  e4 d5 exd5 Qxd5 Nc3 Qe6+ Qe2 Qc6 d3 Bf5 Qf3 Be...   
2556           1309  d4 d5 e3 Nf6 Nf3 b6 Nc3 Bb7 Bb5+ Nc6 Ne5 e6 O-...   
...             ...                                                ...   
20013          2228  e4 e5 Nf3 Nc6 Bb5 g6 d4 exd4 Bxc6 dxc6 Qxd4 Qx...   
20014          2220  e4 e5 Nf3 Nc6 Bc4 Bc5 c3 Nf6 d4 exd4 cxd4 Bb4+...   
20015          2214  d4 Nf6 c4 e6 Nf3 Bb4+ Bd2 a5 e3 O-O Bd3 b6 O-O...   
20016          2148  e4 g6 Nc3 Bg7 f4 c6 g3 d6 Bg2 e5 fxe5 dxe5 d3 ...   
20017          2254  e4 e5 Nf3 Nc6 Bc4 Nf6 Ng5 d5 exd5 Na5 Bb5+ c6 ...   

      opening_eco                                       opening_name  \
950           A08             King's Indian Attack: French Variation   
1794          A29  English Opening: King's English Variation |  F...   
1830          C00                    French Defense: Steinitz Attack   
2554          B01      Scandinavian Defense: Mieses-Kotroc Variation   
2556          D04                    Queen's Pawn Game: Colle System   
...           ...                                                ...   
20013         C60                      Ruy Lopez: Fianchetto Defense   
20014         C54  Italian Game: Classical Variation |  Greco Gam...   
20015         E11        Bogo-Indian Defense: Wade-Smyslov Variation   
20016         B06                                     Modern Defense   
20017         C59  Italian Game: Two Knights Defense |  Steinitz ...   

       opening_ply  
950              6  
1794             7  
1830             3  
2554             4  
2556             5  
...            ...  
20013            6  
20014           17  
20015            8  
20016            2  
20017           17  

[429 rows x 16 columns]

In [4]:
# checking that they are true duplicates:
for id in duplicate_games['id']:
    print(games_df[games_df['id']==id])    

             id  rated    created_at  last_move_at  turns victory_status  \
120    XRuQPSzH   True  1.503580e+12  1.503580e+12     50         resign   
950    XRuQPSzH   True  1.503580e+12  1.503580e+12     50         resign   
12412  XRuQPSzH   True  1.503584e+12  1.503584e+12     50         resign   
13729  XRuQPSzH   True  1.503584e+12  1.503584e+12     50         resign   
14403  XRuQPSzH   True  1.503584e+12  1.503584e+12     50         resign   

      winner increment_code     white_id  white_rating black_id  black_rating  \
120    black           10+0  chess-brahs          2036     cdvh          1706   
950    black           10+0  chess-brahs          2036     cdvh          1706   
12412  black           10+0  chess-brahs          2036     cdvh          1706   
13729  black           10+0  chess-brahs          2036     cdvh          1706   
14403  black           10+0  chess-brahs          2036     cdvh          1706   

                                                   moves

These duplicates are truly duplicates, they can be safely dropped.

In [5]:
# drop duplicates
games_df.drop_duplicates(inplace=True)
games_df.duplicated().sum()

0

With the duplicates gone, the dataset is ready for some EDA.

### Enhance the data

Let's begin by adding some columns:

To enhance the data:

* **"rating_difference"** (int) - the difference in rating between players, using "white_rating" and "black_rating"
* **"avg_rating"** (float) - the average rating between players, using "white_rating" and "black_rating"

To better group the data:

* **"opening"** (string) - a more generalized opening, using "opening_name"
* **"skill_level"** (string) - the category of the players' rating ("beginner", "intermediate", "advanced", "expert, "master"), using "avg_rating"
* **"time_control"** (string) - the category of time increment ("bullet", "blitz", "rapid", "classic"), using "increment_code"

In [6]:
# add rating_difference column by determining difference between white's and black's ratings
games_df['rating_difference'] = games_df['white_rating'] - games_df['black_rating']

# add avg_rating column by determining average between white's and black's ratings
games_df['avg_rating'] = (games_df['white_rating'] + games_df['black_rating']) / 2

# define a function to reduce the complexity of the opening name
def get_opening(opening_name):
    ''' This function will take a game instance from the games_df dataframe and return a shortened version of the opening_name
    
        parameters:
        opening_name - a row from the games_df dataframe

        returns:
        opening (string) - the generalized name of the opening played
        '''    
    return opening_name.split(':')[0]

# define a function to determine skill_level by grouping rating scores into categories
def get_skill_level(rating):
    ''' This function will take a game instance from the games_df dataframe and return a classifying string
    
        parameters:
        game - a row from the games_df dataframe
        
        returns:
        skill_level (string) - the rating classification for players in the game
    '''
    if rating <= 1400:
        return 'beginner'
    elif 1400 < rating <= 1600:
        return 'intermediate'
    elif 1600 < rating <= 1800:
        return 'advanced'
    else:
        return 'master'

# define a function to determine time control by grouping time increments into categories
def get_time_control(increment):
    ''' This function will take a game instance from the games_df dataframe and return a classifying string
    
        parameters:
        game - a row from the games_df dataframe
        
        returns:
        time_control (string) - the time control type for the game
    '''
    time = int(increment.split('+')[0])

    if time < 5:
        return 'bullet'
    elif 5 <= time < 10:
        return 'blitz'
    elif 10 <= time < 30:
        return 'rapid'
    else:
        return 'classic'
    
# create new columns by applying the above functions    
games_df['opening'] = games_df['opening_name'].apply(get_opening)
games_df['skill_level'] = games_df['avg_rating'].apply(get_skill_level)
games_df['time_control'] = games_df['increment_code'].apply(get_time_control)

### Exploring the data

In [7]:
st.header('Total turns played in a game vs. Average rating of the players')

# generate a scatter plot of average player rating vs. total turns, colored by time control
px.scatter(games_df, x='avg_rating', y='turns', color='time_control',  title='Chess Game Total Turns vs. Average Player Rating',
    labels=dict(avg_rating='Average Player Rating', turns='Total Turns', time_control='Time Control'), opacity=.75)

2022-11-03 12:32:09.495 
  command:

    streamlit run C:\Users\joshg\AppData\Roaming\Python\Python310\site-packages\ipykernel_launcher.py [ARGUMENTS]


In [8]:
# observe most frequent openings
top_openings = games_df['opening'].value_counts()[:20]
print(top_openings)

Sicilian Defense           2527
French Defense             1273
Queen's Pawn Game          1039
Italian Game                953
King's Pawn Game            899
Ruy Lopez                   833
English Opening             715
Scandinavian Defense        707
Caro-Kann Defense           581
Scotch Game                 455
Queen's Gambit Declined     384
Van't Kruijs Opening        363
Four Knights Game           353
Zukertort Opening           324
Bishop's Opening            322
Indian Game                 316
Philidor Defense            267
Queen's Gambit Refused      248
Queen's Gambit Accepted     248
Russian Game                247
Name: opening, dtype: int64


In [9]:
st.header('Game End Conditions')

# show histogram of game outcomes, colored by skill level
px.histogram(games_df, x='victory_status', color='skill_level', title='Game End Conditions',
    labels=dict(victory_status='End Condition', skill_level='Skill Level', time_control='Time Control'))


In [10]:
st.header('Rating difference')
st.write('(positive values indicate white is higher rated)')

# create a checkbox to allow the use filter for games where 'rated' is True
rated_only = st.checkbox("Include only rated games")

# create a copy of our data that we can filter if the user chooses to do so
filter_df = games_df
if rated_only:
    filter_df = games_df[games_df['rated']]

# generate a histogram of the rating difference on the filtered data
fig = px.histogram(filter_df, x='rating_difference', nbins=250, range_x=[-500,500])

# use streamlit to display fig
st.write(fig)
fig

In [11]:
st.header('View winner color frequency by opening and skill level')

# generate a list of the 20 most popular openings
top_openings = games_df['opening'].value_counts()[:20]
openings = list(top_openings.index.unique())

# create an enhanced list with a 'Select all' option
openings_2 = openings[:]
openings_2.append('Select all')

# allow the user to select an opening, defaults on 'Select all'
opening = st.selectbox("Select an opening", openings_2, openings_2.index('Select all'))

# if the user choose 'Select all', we need to pass the whole list to our filter
if opening == 'Select all':
    opening = openings

# get a list of the skill levels available to choose from
skill_levels = list(games_df['skill_level'].unique())

# create an enhanced list with a 'Select all' option
skill_levels_2 = skill_levels[:]
skill_levels_2.append('Select all')

# allow the user to select a skill level, defaults on 'Select all'
skill_level = st.selectbox("Select a skill level", skill_levels_2, skill_levels_2.index('Select all'))

# if the user chooses 'Select all', we need to pass the whole list to our filter
if skill_level == 'Select all':
    skill_level = skill_levels

# filter our data per users choices
q_string = 'opening in @opening and skill_level in @skill_level'
filter_df = games_df.query(q_string)

# generate the histogram of winner color, in a consistent order with plotly express
fig = px.histogram(filter_df, x='winner', category_orders={'winner':['white','black','draw']})

# display the figure with streamlit
st.write(fig)
fig